vpr/src/place/analytic_placer.cpp

```diff
@@ -0,0 +1,705 @@
+    ymin = std::max<int>(ymin, 1);
+    xmax = std::max<int>(xmax, 1);
+    ymax = std::max<int>(ymax, 1);
+
+    return (ymax - ymin) + (xmax - xmin);
+}
+
+wirelen_t AnalyticPlacer::total_hpwl() {
+    const ClusteredNetlist& clb_nlist = g_vpr_ctx.clustering().clb_nlist;
+
+    wirelen_t hpwl = 0;
+    for (auto net_id : clb_nlist.nets()) {
+        if (!clb_nlist.net_is_ignored(net_id)) {
+            hpwl += get_net_hpwl(net_id);
+        }
+    }
+    return hpwl;
+}
+
+// Setup the cells to be solved, returns the number of rows (number of blks to solve)
+int AnalyticPlacer::setup_solve_blks(t_logical_block_type_ptr blkTypes) {
+    const ClusteredNetlist& clb_nlist = g_vpr_ctx.clustering().clb_nlist;
+    PlacementContext& place_ctx = g_vpr_ctx.mutable_placement();
+
+    int row = 0;
+    solve_blks.clear();
+    // clear udata of all cells
+    for (auto& blk : blk_row_num) {
+        blk.second = dont_solve;
+    }
+    // update cells to be placed, excluding cell children
+    for (auto blk_id : place_blks) {
+        if (blkTypes != (clb_nlist.block_type(blk_id)))
+            continue;
+        blk_row_num[blk_id] = row++;
+        solve_blks.push_back(blk_id);
+    }
+    // update udata of children
+    for (auto& macro : place_ctx.pl_macros)
+        for (auto& member : macro.members)
+            blk_row_num[member.blk_index] = blk_row_num[macro_head(member.blk_index)];
+
+    return row;
+}
+
+// Update the location of all members of all macros
+void AnalyticPlacer::update_macros() {
+    for (auto& macro : g_vpr_ctx.mutable_placement().pl_macros) {
+        ClusterBlockId head_id = macro.members[0].blk_index;
+        for (auto member = ++macro.members.begin(); member != macro.members.end(); ++member) {
+            blk_locs[member->blk_index].loc = blk_locs[head_id].loc + member->offset;
+        }
+    }
+}
+
+// Build and solve in one direction
+void AnalyticPlacer::build_solve_direction(bool yaxis, int iter, int build_solve_iter) {
+    for (int i = 0; i < build_solve_iter; i++) {
+        EquationSystem<double> esx(solve_blks.size(), solve_blks.size());
+        build_equations(esx, yaxis, iter);
+        solve_equations(esx, yaxis);
+    }
+}
+
+// Build the system of equations for either X or Y
+void AnalyticPlacer::build_equations(EquationSystem<double>& es, bool yaxis, int iter) {
+    const ClusteredNetlist& clb_nlist = g_vpr_ctx.clustering().clb_nlist;
+    PlacementContext& place_ctx = g_vpr_ctx.mutable_placement();
+
+    // Return the x or y position of a block
+    auto blk_p = [&](ClusterBlockId blk_id) { return yaxis ? blk_locs.at(blk_id).loc.y : blk_locs.at(blk_id).loc.x; };
+    // Return legal position from legalization, after first iteration
+    auto legal_p = [&](ClusterBlockId blk_id) { return yaxis ? blk_locs.at(blk_id).legal_loc.y : blk_locs.at(blk_id).legal_loc.x; };
+    es.reset();
+
+    for (auto net_id : clb_nlist.nets()) {
+        if (clb_nlist.net_is_ignored(net_id)) continue;
```

Comment:
I don't like putting the logic on the same line as the if; please put on the line underneath, indented.